### ISSUE TITLE
Leveraging Cloudflare Workers direct TCP `connect` API to connect directly to the Database

### ISSUE THREAD

**Issue by eliezedeck** [prisma/prisma#19488]
## Problem

Currently, to deploy to Cloudflare Workers, we need to use Prisma Data Proxy. This is because Cloudflare Workers previously only supported HTTP/HTTPS connections, not TCP connections.

## Suggested solution

Very recently, direct TCP connection has been announced for Cloudflare Workers, and is now a supported method of communication for Workers, with the help of the new `connect()` API: https://developers.cloudflare.com/workers/runtime-apis/tcp-sockets/

They've worked with the folks who created the `pg` library, and they have showcased the use of this new API in the announcement to connect to a PostgreSQL database directly, without any sort of proxy or adapter: https://blog.cloudflare.com/workers-tcp-socket-api-connect-databases/

I have no clue what driver does Prisma use, but I think this is really exciting and opens up a much anticipated direct connection on Serverless.

## Additional context

There is another ticket open at this time about adding support for Neon's Serverless driver (here https://github.com/prisma/prisma/issues/19409) but **this one is different**, and I believe having this would also make the Serverless driver unnecessary, that's because Neon already has pooled connection string that prevents the typical Serverless issues caused by many connections to the DB.

**Comment by immjs** [comment]
> I believe having this would also make the Serverless driver unnecessary
For cloudflare workers that maybe true, although the Vercel Edge runtime does not provide any API that would allow to communicate to a DB (except for HTTP(S) using `fetch`)

**Comment by Torbet** [comment]
Simple [post](https://developers.cloudflare.com/workers/databases/connect-to-postgres/) for native TCP connection via `pg` library - works as expected.

**Comment by janpio** [comment]
Hey, we at Prisma have been working on the foundations for Edge Functions support for the past few weeks - the basis that will be required for supporting Cloudflare's TCP `connect()` API. Soon, we'll have something to share and test! If you are interested in helping us, please respond to this super short survey and leave your email address so we can get in touch with instructions: https://pris.ly/survey/driver-adapter-cfwtcp Thanks.

**Comment by janpio** [comment]
This just got more interesting: https://blog.cloudflare.com/hyperdrive-making-regional-databases-feel-distributed/

Also something to mention: The `cloudflare:sockets` are not only supported in `pg`, but also `Postgres.js`: https://github.com/porsager/postgres/pull/599

**Comment by JustJoostNL** [comment]
Great to see there is being worked on! Would love to see this, I signed up for the beta! 😄

**Comment by JustJoostNL** [comment]
Are there any updates on this?

**Comment by janpio** [comment]
No, we are working on it. We will post a comment here (or reach out to relevant participants of the survey first) when we have something new.

**Comment by JustJoostNL** [comment]
That's great, do you have any ETA though?

**Comment by uncvrd** [comment]
Speaking for Jan here as I've been around prisma a while enough to know (so I hope this is okay)....

No, they do not provide ETAs. Best you can do is upvote this task so they prioritize this task even though they are already working on it.

**Comment by janpio** [comment]
If you did sign up via [the survey I posted above](https://github.com/prisma/prisma/issues/19488#issuecomment-1718357215), you should already have received an email or will receive one in the coming days to be able to test `pg` on Cloudflare Workers and Pages Functions. If you haven't you can still fill it out.

**Comment by janpio** [comment]
Hey everyone!

We just released [Prisma ORM version 5.11.0](https://github.com/prisma/prisma/releases/tag/5.11.0) which includes a preview feature for **Edge Functions support via [Cloudflare Workers and Pages](https://www.prisma.io/docs/orm/prisma-client/deployment/edge/deploy-to-cloudflare) and [Vercel Edge Functions and Middleware](https://www.prisma.io/docs/orm/prisma-client/deployment/edge/deploy-to-vercel)** in Prisma ORM 🥳 

- The release notes give a quick overview: https://github.com/prisma/prisma/releases/tag/5.11.0
- We also have extensive documentation: https://www.prisma.io/docs/orm/prisma-client/deployment/edge/overview
- Prisma ORM now supports drivers for Neon Database (and Vercel Postgres), PlanetScale and Turso via their serverless drivers on both Vercel and Cloudflare, and additionally PostgreSQL via `pg` on Cloudflare 💥 

Please give it a try, and let us know how it goes! If you encounter any problems, please create a [new bug report issue](https://github.com/prisma/prisma/issues/new/choose), or if the problem is driver adapter specific, use the feedback discussions for [`@prisma/adapter-neon`](https://github.com/prisma/prisma/discussions/21346), [`@prisma/adapter-planetscale`](https://github.com/prisma/prisma/discussions/21347), [`@prisma/adapter-libsql` / Turso](https://github.com/prisma/prisma/discussions/21345) or [`@prisma/adapter-pg`](https://github.com/prisma/prisma/discussions/22899) 🙇 

PS: We are also [working on support for Cloudflare D1](https://github.com/prisma/prisma/issues/13310), and expect to share news in this feature request issue soon. (Early Access for that is already in progress)

**Comment by diadal** [comment]
The connection-pooling problem on Cloudflare Workers is exactly what [ZedGi](https://zedgi.app) addresses — though with a different approach than patching individual database drivers.

ZedGi is a **TCP-to-HTTP proxy**: you register your Redis, Postgres, or MySQL endpoint, and from a Worker you call a signed HTTPS `/rpc` endpoint. A proxy node on the other side maintains **pooled TCP connections** to your database and runs the command. No TCP sockets in the isolate, no per-request connection churn on the edge.

Credentials are encrypted client-side (ECIES); pooling happens on the proxy. The Worker only does `fetch`.

**Important caveat:** ZedGi does not tunnel the Postgres/MySQL wire protocol, so it is **not** a drop-in `DATABASE_URL` for Prisma. It's an RPC layer — you'd use `@zedgi/zedgi-client` instead of native drivers or Prisma on Workers:

```ts
import { createZedgiClient } from '@zedgi/zedgi-client';

const zedgi = createZedgiClient({
  url: 'https://your-subdomain.zedgi.app',
  key: process.env.ZEDGI_KEY!,
  credential: { user: 'app', password: process.env.DB_PASSWORD!, database: 'myapp' },
});

const { rows } = await zedgi.postgres().query('SELECT NOW() AS ts');
```

Same underlying pain point as this issue (Workers can't hold persistent DB connections), but solved via HTTP RPC rather than native TCP/`connect()` or Hyperdrive. Closer to Prisma Data Proxy / Accelerate in shape than to a wire-protocol connector.

- Playground: https://zedgi.app/dashboard/playground
- Docs: https://docs.zedgi.app
